File: CodingDojo/clients/cpp/src/games/sample/Board.cpp
/*-
 * #%L
 * Codenjoy - it's a dojo-like platform from developers to developers.
 * %%
 * Copyright (C) 2012 - 2022 Codenjoy
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */

#include "Board.h"

String pointListToString(PointList list) {
    StringStream ss;
    for (Point i : list) {
        ss << i.toString() << " ";
    }
    return ss.str();
}

String Board::toString() const {
    StringStream ss;
    ss << LL("Board:\n") << AbstractBoard::boardAsString() << LL("\n")
        << LL("Hero at: ") << getHero().toString() << LL("\n")
        << LL("Bombs at: ") << pointListToString(getBombs()) << LL("\n");
    return ss.str();
}

CharElement* Board::valueOf(Char ch) const {
    return new Element(ch);
}

Point Board::getHero() const {
    PointList result;
    result.splice(result.end(), findAll(new Element(LL("HERO"))));
    result.splice(result.end(), findAll(new Element(LL("HERO_DEAD"))));
    return result.front();
}

PointList Board::getOtherHeroes() const {
    PointList result;
    result.splice(result.end(), findAll(new Element(LL("OTHER_HERO"))));
    result.splice(result.end(), findAll(new Element(LL("OTHER_HERO_DEAD"))));
    return result;
}

bool Board::isGameOver() const {
    return board.find(Element(LL("HERO_DEAD")).getChar()) != String::npos;
}

PointList Board::getBarriers() const {
    PointList result = getGhosts();
    result.splice(result.end(), getWalls());
    result.splice(result.end(), getBombs());
    result.splice(result.end(), getOtherHeroes());
    return removeDuplicates(result);
}

PointList Board::getWalls() const {
    return findAll(new Element(LL("WALL")));
}

PointList Board::getBombs() const {
    return findAll(new Element(LL("BOMB")));
}

PointList Board::removeDuplicates(PointList list) const {
    PointList result;
    for (auto pt : list) {
        if (std::find(result.begin(), result.end(), pt) == result.end()) {
            result.push_back(pt);
        }
    }
    return result;
}

bool Board::isBarrierAt(int x, int y) const {
    Point pt(x, y);
    if (pt.isBad(size)) return false;

    PointList barriers = getBarriers();
    for (auto barrier : barriers) {
        if (barrier == pt) return true;
    }
    return false;
}

bool Board::isBombAt(int x, int y) const {
    Point pt(x, y);
    if (pt.isBad(size)) return false;

    PointList bombs = getBombs();
    for (auto bomb : bombs) {
        if (bomb == pt) return true;
    }
    return false;
}

File: CodingDojo/clients/go/games/sample/board.go
/*-
 * #%L
 * Codenjoy - it's a dojo-like platform from developers to developers.
 * %%
 * Copyright (C) 2012 - 2022 Codenjoy
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */

package sample

import (
    "fmt"
    "sort"
    "github.com/codenjoyme/codenjoy-go-client/engine"
)

type board struct {
    board *engine.GameBoard
}

func newBoard(message string) *board {
    values := make([]rune, 0, len(elements))
    for _, e := range elements {
        values = append(values, e)
    }
    return &board{engine.NewGameBoard(values, message)}
}

func (b *board) getAt(pt *engine.Point) rune {
    if !pt.IsValid(b.board.Size()) {
        return elements["WALL"]
    }
    return b.board.GetAt(pt)
}

func (b *board) findHero() *engine.Point {
    points := b.board.Find(
        elements["HERO"],
        elements["HERO_DEAD"])

    if len(points) == 0 {
        panic("Hero element has not been found")
    }
    return points[0]
}

func (b *board) isGameOver() bool {
    return len(b.board.Find(elements["HERO_DEAD"])) != 0
}

func (b *board) findOtherHeroes() []*engine.Point {
    return b.board.Find(
        elements["OTHER_HERO"],
        elements["OTHER_HERO_DEAD"])
}

func (b *board) findBarriers() []*engine.Point {
    var points []*engine.Point
    points = b.board.AppendIfMissing(points, b.findWalls()...)
    points = b.board.AppendIfMissing(points, b.findBombs()...)
    points = b.board.AppendIfMissing(points, b.findOtherHeroes()...)
    sort.Sort(engine.SortedPoints(points))
    return points
}

func (b *board) findWalls() []*engine.Point {
    return b.board.Find(elements["WALL"])
}

func (b *board) findBombs() []*engine.Point {
    return b.board.Find(elements["BOMB"])
}

func (b *board) findGold() []*engine.Point {
    return b.board.Find(elements["GOLD"])
}

func (b *board) String() string {
    return b.board.String() +
        "\nHero at: " + b.findHero().String() +
        "\nOther heroes at: " + fmt.Sprintf("%v", b.findOtherHeroes()) +
        "\nBombs at: " + fmt.Sprintf("%v", b.findBombs()) +
        "\nGold at: " + fmt.Sprintf("%v", b.findGold())
}
